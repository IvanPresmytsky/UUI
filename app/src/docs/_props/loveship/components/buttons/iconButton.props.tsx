import * as React from 'react';
import { IconButton, IconButtonProps } from '@epam/loveship';
import { ColorPicker, DocBuilder } from '@epam/uui-docs';
import { onClickDoc, isDisabledDoc, iCanRedirectDoc, iconDoc, DefaultContext, FormContext } from '../../docs';
import { colors } from '../../docs/helpers/colorMap';

const IconButtonDoc = new DocBuilder<IconButtonProps>({ name: 'IconButton', component: IconButton })
    .implements([
        onClickDoc, isDisabledDoc, iCanRedirectDoc, iconDoc,
    ])
    .prop('color', {
        renderEditor: (editable, examples) => <ColorPicker colors={ examples.map((i) => ({ value: i, hex: colors[i] })) } { ...editable } />,
        examples: ['sky', 'grass', 'sun', 'fire', 'cobalt', 'violet', 'fuchsia', 'white', 'night500', 'night600'],
    })
    .withContexts(DefaultContext, FormContext);

export default IconButtonDoc;
