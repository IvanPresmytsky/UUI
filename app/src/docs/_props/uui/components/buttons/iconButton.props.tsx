import * as React from 'react';
import { allIconColors, IconButton, IconButtonProps } from '@epam/uui';
import { ColorPicker, DocBuilder } from '@epam/uui-docs';
import { onClickDoc, isDisabledDoc, iCanRedirectDoc, iconDoc } from '../../docs';
import { DefaultContext } from '../../docs';

const IconButtonDoc = new DocBuilder<IconButtonProps>({ name: 'IconButton', component: IconButton })
    .implements([
        onClickDoc, isDisabledDoc, iCanRedirectDoc, iconDoc,
    ])
    .prop('color', { renderEditor: (editable: any, examples) => <ColorPicker colors={ examples.map((i) => ({ value: i })) } { ...editable } />, examples: allIconColors })
    .withContexts(DefaultContext);

export default IconButtonDoc;
