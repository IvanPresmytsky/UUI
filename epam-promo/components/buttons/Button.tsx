import * as uui from '@epam/uui';
import { createSkinComponent, devLogger } from '@epam/uui-core';
import { FillStyle } from '../types';
import css from './Button.module.scss';

export type ButtonColor = 'blue' | 'green' | 'red' | 'gray50' | 'gray';

export interface ButtonMods {
    /** @default 'solid' */
    fill?: FillStyle;
    /** @default 'blue' */
    color?: ButtonColor;
    /** @default '36' */
    size?: uui.ButtonMods['size'];
}

const mapFill: Record<FillStyle, uui.ButtonFill> = {
    solid: 'solid',
    white: 'outline',
    light: 'ghost',
    none: 'none',
};

export type ButtonProps = uui.ButtonCoreProps & ButtonMods;

export const Button = createSkinComponent<uui.ButtonProps, ButtonProps>(
    uui.Button as any, // TODO: remove it when BaseButton inheritance will be reworked
    (props) => {
        if (__DEV__) {
            devLogger.warnAboutDeprecatedPropValue<ButtonProps, 'color'>({
                component: 'Button',
                propName: 'color',
                propValue: props.color,
                propValueUseInstead: 'gray',
                condition: () => ['gray50'].indexOf(props.color) !== -1,
            });
        }
        return {
            fill: mapFill[props.fill] || mapFill.solid,
        } as any; // TODO: need new helper to rewrite types
    },
    (props) => [
        ['42', '48'].includes(props.size) && css.uppercase,
    ],
);
