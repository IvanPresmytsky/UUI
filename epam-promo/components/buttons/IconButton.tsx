import { createSkinComponent } from '@epam/uui-core';
import { EpamPrimaryColor } from '../types';
import * as uui from '@epam/uui';

export type IconColor = EpamPrimaryColor | 'gray30' | 'gray50' | 'gray60';

export interface IconButtonMods {
    /**
     * @default 'gray60'
     */
    color?: IconColor;
}

export type IconButtonProps = uui.IconButtonCoreProps & IconButtonMods;

export const IconButton = createSkinComponent<uui.IconButtonProps, IconButtonProps>(
    uui.IconButton,
    (props) => ({ color: props.color ?? 'gray60' }),
);
