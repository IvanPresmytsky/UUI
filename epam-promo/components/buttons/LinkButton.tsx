import { createSkinComponent, devLogger } from '@epam/uui-core';
import { LinkButtonCoreProps, LinkButton as UuiLinkButton, LinkButtonProps as UuiLinkButtonProps } from '@epam/uui';

export interface LinkButtonMods {
    /**
     * @default 'blue'
     */
    color?: 'blue' | 'green' | 'amber' | 'red' | 'gray60' | 'gray10';
}

export type LinkButtonProps = LinkButtonCoreProps & LinkButtonMods;

export const LinkButton = createSkinComponent<UuiLinkButtonProps, LinkButtonProps>(
    UuiLinkButton,
    (props) => {
        if (__DEV__) {
            devLogger.warnAboutDeprecatedPropValue<LinkButtonProps, 'color'>({
                component: 'LinkButton',
                propName: 'color',
                propValue: props.color,
                condition: () => ['green', 'amber', 'red'].indexOf(props.color) !== -1,
            });
        }
        return {
            color: props.color ?? 'blue',
        } as LinkButtonProps;
    },
    () => [],
);
