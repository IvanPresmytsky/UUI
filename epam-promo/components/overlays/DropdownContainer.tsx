import { createSkinComponent } from '@epam/uui-core';
import * as uui from '@epam/uui';

type DropdownContainerMods = {
    /**
     * Defines component color.
     * @default 'white'
     * */
    color?: 'white' | 'gray70';
};

/** Represents the properties of a DropdownContainer component. */
export type DropdownContainerProps = uui.DropdownContainerProps & DropdownContainerMods;

function applyDropdownContainerMods(mods: DropdownContainerMods) {
    return [`uui-color-${mods.color || 'white'}`];
}

export const DropdownContainer = createSkinComponent<uui.DropdownContainerProps, DropdownContainerProps>(
    uui.DropdownContainer,
    () => null,
    applyDropdownContainerMods,
);
