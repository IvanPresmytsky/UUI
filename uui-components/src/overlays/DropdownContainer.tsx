import * as React from 'react';
import FocusLock from 'react-focus-lock';
import {
    uuiElement, IHasCX, IHasChildren, cx, IHasRawProps, uuiMarkers, IHasForwardedRef, IDropdownBodyProps,
} from '@epam/uui-core';
import { VPanel } from '../layout/flexItems/VPanel';
import PopoverArrow from './PopoverArrow';

type FocusLockProps = Pick<React.ComponentProps<typeof FocusLock>, 'returnFocus' | 'persistentFocus' | 'lockProps' | 'shards' | 'as' | 'className'>;
export interface DropdownContainerProps
    extends IHasCX,
    IHasChildren,
    IHasRawProps<React.HTMLAttributes<HTMLDivElement>>,
    IHasForwardedRef<HTMLDivElement>,
    IDropdownBodyProps,
    FocusLockProps {
    width?: number | 'auto';
    maxWidth?: number;
    /**
     * Wraps DropdownContainer with FocusLock component to support keyboard navigation. It's `true` by default.
     * After DropdownContainer appeared the focus will be set on the first focusable element inside.
     * 
     * You can also specify the following props for FocusLock: 'returnFocus', 'persistentFocus', 'lockProps', 'shards', 'as', 'className' by passing them as DropdownContainer's props.
     * By default 'returnFocus' and 'persistentFocus' are true. It means that the focus is locked inside the component and the focus returns into initial position on unmount.
     */
    focusLock?: boolean;
    /**
     * If `true` it handles Escape key press on FocusLock wrappers (see `focusLock` prop) and calls `props.onClose()`.
     * By default the value is `true`.
     */
    closeOnEsc?: boolean;
    height?: number;
    showArrow?: boolean;
    style?: React.CSSProperties;
}

class DropdownContainerInner extends React.Component<DropdownContainerProps> {
    render() {
        return (
            <VPanel
                forwardedRef={ this.props.forwardedRef }
                cx={ cx(uuiElement.dropdownBody, this.props.cx, uuiMarkers.lockFocus) }
                style={ {
                    ...this.props.style, minWidth: this.props.width, minHeight: this.props.height, maxWidth: this.props.maxWidth,
                } }
                rawProps={ { tabIndex: -1, ...this.props.rawProps } }
            >
                {this.props.children}
                {this.props.showArrow && (
                    <PopoverArrow ref={ this.props?.arrowProps?.ref } arrowProps={ this.props?.arrowProps } placement={ this.props?.placement || 'bottom-start' } />
                )}
            </VPanel>
        );
    }
}

export function DropdownContainer(props: DropdownContainerProps) {
    const handleEscape = (e: React.KeyboardEvent<HTMLElement>) => {
        if (e.key === 'Escape' && props.isOpen) {
            props.onClose?.();
        }
    };

    const {
        focusLock,
        returnFocus,
        persistentFocus,
        closeOnEsc,
        lockProps,
        shards,
        as,
        className,
    } = props;

    const focusLockProps = {
        returnFocus,
        persistentFocus,
        lockProps: {
            ...(closeOnEsc && { onKeyDown: handleEscape }),
            ...lockProps,
        },
        ...(shards && { shards }),
        ...(as && { as }),
        className,
    };

    return focusLock
        ? (
            <FocusLock { ...focusLockProps } ref={ lockProps.lockRef }>
                <DropdownContainerInner { ...props }>{props.children}</DropdownContainerInner>
            </FocusLock>
        )
        : (<DropdownContainerInner { ...props }>{props.children}</DropdownContainerInner>);
}

DropdownContainer.defaultProps = {
    focusLock: true,
    closeOnEsc: true,
    returnFocus: true,
    persistentFocus: true,
};
