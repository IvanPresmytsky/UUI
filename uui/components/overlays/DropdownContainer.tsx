import { withMods } from '@epam/uui-core';
import * as uuiComponents from '@epam/uui-components';
import css from './DropdownContainer.module.scss';

type DropdownContainerMods = {
    /**
     * Defines vertical padding.
     */
    vPadding?: '6' | '12' | '18' | '24' | '30' | '48';
    /**
     * Defines horizontal padding.
     */
    padding?: '6' | '12' | '18' | '24' | '30';
};

/** Represents the properties of a DropdownContainer component. */
export type DropdownContainerProps = uuiComponents.DropdownContainerProps & DropdownContainerMods;

function applyDropdownContainerMods(mods: DropdownContainerMods) {
    return [
        css.root,
        mods.vPadding && `vPadding-${mods.vPadding}`,
        mods.padding && `padding-${mods.padding}`,
    ];
}

export const DropdownContainer = withMods<uuiComponents.DropdownContainerProps, DropdownContainerMods>(uuiComponents.DropdownContainer, applyDropdownContainerMods);
