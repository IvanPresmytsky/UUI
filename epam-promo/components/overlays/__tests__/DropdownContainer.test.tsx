import React from 'react';
import { DropdownContainer } from '../DropdownContainer';
import { renderer } from '@epam/uui-test-utils';
import { Button } from '../../buttons';

describe('DropdownContainer', () => {
    it('should be rendered correctly', () => {
        const tree = renderer
            .create(
                <DropdownContainer focusLock={ false }>
                    <Button />
                </DropdownContainer>,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly', () => {
        const tree = renderer
            .create(
                <DropdownContainer color="gray70" width={ 300 } height={ 600 } focusLock={ false }>
                    <Button />
                </DropdownContainer>,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
