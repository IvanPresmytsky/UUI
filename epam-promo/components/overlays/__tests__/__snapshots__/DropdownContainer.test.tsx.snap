// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownContainer should be rendered correctly 1`] = `
Array [
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={0}
  />,
  <div
    data-focus-lock-disabled={false}
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
  >
    <div
      className="uui-dropdown-body -lock-focus root dropdown-container-white container"
      style={
        Object {
          "maxWidth": undefined,
          "minHeight": undefined,
          "minWidth": undefined,
        }
      }
      tabIndex={-1}
    >
      <button
        className="container uui-button-box uui-enabled button-primary root size-36 mode-solid"
        onClick={[Function]}
        tabIndex={-1}
        type="button"
      />
    </div>
  </div>,
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={0}
  />,
]
`;

exports[`DropdownContainer should be rendered correctly 2`] = `
<div
  className="uui-dropdown-body -lock-focus root dropdown-container-gray70 container"
  style={
    Object {
      "maxWidth": undefined,
      "minHeight": 600,
      "minWidth": 300,
    }
  }
  tabIndex={-1}
>
  <button
    className="container uui-button-box uui-enabled button-primary root size-36 mode-solid"
    onClick={[Function]}
    tabIndex={-1}
    type="button"
  />
</div>
`;
